### ISSUE TITLE
Commit 52d57f9c7 broke my key-reading code

### ISSUE THREAD

**Issue by kurtdekker** [pygame/pygame#3818]
For 15+ years my PyGame wrapper has done:

	keysprevious = list( keysnow)
	keysnow = pygame.key.get_pressed()

in one place to simultaneously age and collect key input.

Commit 52d57f9c7 breaks that.

Now I get:

	TypeError: Iterating over key states is not supported

when I invoke the list() ctor

What was the point of that commit?

Does it bring any value to an already-amazing library?

What problem does it solve exactly?

What changed after all this time to necessitate such a commit?

**Comment by kurtdekker** [comment]
Forgot to add that the commit message itself seems to be of no help:

    The collection of bools returned by ``get_pressed`` can not be iterated
    over because the indexes of the internal tuple does not correpsond to the 
    keycodes.

Because numeric correspondence is not an (obvious) prerequisite for iterating something.

And the API definition already specifies you index the array via the K_xxxx constants.

**Comment by illume** [comment]
Ok, seems like we should revert it.

**Comment by Klaus-in-Tech** [comment]
Am also facing the same error yet the resource that am trying to work along is using the exact code and it is working fine mine is throwing that 

``TypeError: Iterating over key states is not supported``

**Comment by nelson19721005** [comment]
I have the same problem.

TypeError: Iterating over key states is not supported

Any help please.

**Comment by Klaus-in-Tech** [comment]
If you look at the documentation for [pygame.key.get_pressed](https://www.pygame.org/docs/ref/key.html#pygame.key.get_pressed)
It indicates iteration isn't possible from PyGame 2.2. You could try an older version of PyGame or a different approach.

**Comment by illume** [comment]
Here is the offending PR that broke things https://github.com/pygame/pygame/pull/3495

**Comment by Dreamsorcerer** [comment]
@kurtdekker My question is what are you managing to use `keysprevious` for?

This will just produce a list of bools with no obvious way to tell which keys are actually pressed (as explained in #3437).
Typically, you want an obvious way to tell which keys are pressed (e.g. space key etc.).

In pygame 1, you could map the bools to the key constants, but in pygame 2 this no longer works (so, ideally, the functionality would never have been included in 2.0, but nobody noticed the issue till I pointed it out after a friend tried to use some pygame 1 tutorial).

**Comment by Dreamsorcerer** [comment]
> Because numeric correspondence is not an (obvious) prerequisite for iterating something.
> 
> And the API definition already specifies you index the array via the K_xxxx constants.

As explained in that issue, K_SPACE is 32, but the key corresponding to that value is at index 44. So, indexing the array with the constants will not work, _because_ numeric correspondence is a prerequisite.

Your comment just suggests that the removing the feature was correct, because it sounds like your code is broken and you don't even know it. Exactly the reason I suggested removing it.

Maybe I'm misunderstanding something, but it still looks like a feature that was broken in Pygame 2.0 and provides no utility now (only confusing users who think it still works like Pygame 1).

**Comment by illume** [comment]
It would be best if we can keep compatibility with pygame 1… somehow.

But it sounds like there’s a few apps that work ok with the code previous to this. That’s preferable IMHO.

**Comment by Dreamsorcerer** [comment]
Yeah, but why do they work? That's what I want to know.

If you can't use it to tell which keys are pressed, then what is it being used for? Counting how many keys are pressed (surely an iterator would be a bad way to do that)?

I think it'd be best to understand how it's being used first, so I'd wait till one of the above commenters can provide more information.

**Comment by illume** [comment]
Yeah, more info would be good. You’re right.

I’d like to know if there’s some way we can get the constants to move back so this works. I dunno if we can hack that in underneath so there’s a translation. Anyway… something to look into.

**Comment by illume** [comment]
I've reverted that change so this works again. It will be out in pygame==2.5.0

```python
keysprevious = list(keysnow)
keysnow = pygame.key.get_pressed()
```

**Comment by Dreamsorcerer** [comment]
Which I still think will just lead to more people having broken code and not knowing why. It's still not clear if any of the people who posted here actually have a functioning game under Pygame 2, and are not just expecting the behaviour will be the same as Pygame 1.

**Comment by illume** [comment]
Like you say, there are some usages like this that are broken:

```python
for key_constant, pressed in enumerate(pygame.key.get_pressed()):
    if pressed:
        key_name = pygame.key.name(key_constant)
```

There are some usages found with this search: https://github.com/search?q=list%28pygame.key.get_pressed%28%29%29&type=code

I would like to have a solution that doesn't break this (from the original poster above):

```python
keysprevious = list(keysnow)
keysnow = pygame.key.get_pressed()
```

Please correct me if I'm wrong... but that continues to work in pygame 2 (not 2.2, 2.3, 2.4)? Some of those other usages look fine as well?

I'd like to see other solutions explored before giving up and removing iter on the get_pressed(). So far I haven't seen the work anyone has done exploring other possible solutions... and finding them inadequate.

**Comment by Dreamsorcerer** [comment]
> Please correct me if I'm wrong... but that continues to work in pygame 2 (not 2.2, 2.3, 2.4)? Some of those other usages look fine as well?

The question is what do they do with `keysnow`? That's very similar to the original code shown in #3437, which is broken. I'm still unsure what a user could be doing with `keysnow` that is useful...

The same question goes to your Github search, I can see that some (if not all) of them are broken code because of the exact same reasons.
e.g. One immediately does `pressed[pygame.K_LCTRL] or pressed[pygame.K_RCTRL]`, which is exactly the kind of code which is broken.

**Comment by illume** [comment]
Probably keeping many of the old codes for keys is possible. The letters are as before...

```
list(pygame.key.get_pressed())[K_a] -> currently works
```

But some keys don't work obs...
```
list(pygame.key.get_pressed())[K_UP] -> currently fails
```

Having a mapping between the old values and newer values might be possible. 


Where we can 
- map the constant values to the old values when giving them **from** SDL
- mapping them back when sending them **to** SDL

SDL2, pygame 2.x
```
>>> pygame.K_UP
1073741906
>>> pygame.K_a
97
```

SDL1, pygame 1.9.6
```
>>> pygame.K_UP
273
>>> pygame.K_a
97
```

So K_UP goes back to 273, but everywhere we send that back to SDL we map it back to 1073741906.


Might be possible. I can't remember if we checked that in the past sometime. Maybe there's clashes, I haven't checked yet.

**Comment by Dreamsorcerer** [comment]
Yes, if it's possible to change the constants to something that matches the indices of the iterator, that could work. I guess the risk with converting them back to SDL is that someone might be saving SDL values or something and not expect them to be converted. I guess we always convert and enforce pygame's constants everywhere, that would work (but, may cause some backwards compatible breakage for those who wrote code with pygame 2.0-2.4).
